### PR TITLE
Improvements to bokehjs' tests framework

### LINKED ISSUE #14036: JS unit tester issues
@mattpap I've run into some issues with the current unit tester:

- [x] `Error: expected 0 to be equal to 0` The tester is distinguishing between -0 and +0 which it should not do for numerical equality comparisons. I've had to resort to clunky `return 0 - padding` instead of `return -padding` to work around this in cases where `padding` is zero. 

- [x] The number of failed tests should be printed out at the very end. Often times if lots of tests are failing I first want a quick check whether a change made *any* difference, and a quick glance at the number of failed tests would provide that. But currently it is hard to find the number of failed tests buried in the middle of long output. 

- [x] The actual number of successful tests should be printed out. Currently shows something like 
   ```
    1934 of 1934 | 1809 skipped | 0s
    ```
    I don't want to do math in my head to figure out that 125 of 125 of actually run tests passed. 

- [ ] The actual code for the assertion that failed should be printed in the failure output, to make it easy to copy and paste and code grep. Failing that, at the very least this last line of the trace is not helpful and makes it harder to find the real location:
   ```
   at Asserts.equal (http://localhost:5777/static/test/@@/build/test/unit/assertions.js:82:1)
   ```
   This should be discarded. All that is truly needed is this part:
   ```
   at http://localhost:5777/static/test/@@/build/test/unit/models/ranges/factor_range.js:1112:1
   ```
   everything else in the trace is just noise that explodes the output size. 

   EDIT: and I just noticed the line number is into the generated JS too, not into the original test file. In general it is much harder to back out from an error to a line in a test file that in should be. 

- [ ] Mysterious error reporting in some cases, see https://github.com/bokeh/bokeh/pull/14037#discussion_r1730827587

### PR DESCRIPTION
**Pull request by mattpap**
```
[01:24:04] Starting 'test:integration'...
Running in Chrome/118.0.5993.88 using devtools protocol 1.3
████████████████████████████████████████ 100% | 6 of 6 | 6 failures | 1s

✗ gridplot()
 └── should align axes when toolbar_location=null
images differ by 2px (0.00%)

✗ gridplot()
 └── should align axes when toolbar_location=above
images differ by 2px (0.00%)

✗ gridplot()
 └── should align axes when toolbar_location=right
images differ by 2px (0.00%)

✗ gridplot()
 └── should align axes when toolbar_location=left
images differ by 2px (0.00%)

✗ gridplot()
 └── should align axes when toolbar_location=below
images differ by 2px (0.00%)

✗ Bug
 └┬─ in issue #11035
  └── doesn't allow to use non-Plot models in gridplot()
images differ by 2px (0.00%)

6 of 6 selected tests failed (out of 1173 total tests)
[01:24:06] Finished 'test:integration' after 2.84 s
```

addresses second and third point in #14036

### CODE REVIEW

File: bokehjs/test/devtools/devtools.ts [hunk 1]
```diff
@@ -317,40 +317,40 @@ async function run_tests(ctx: TestRunContext): Promise<boolean> {
...
         }
       }
 
-      if (test_suite.length == 0) {
-        fail("nothing to test")
-      }
+      const test_suite = all_tests.filter(([, test]) => test.omit !== true)
+
+      const num_all_tests = all_tests.length
+      const num_test_suite = test_suite.length
```
Comment (hoxbro) [inline bokehjs/test/devtools/devtools.ts]:
```suggestion
      const selected_tests = all_tests.filter(([, test]) => test.omit !== true)

      const num_all_tests = all_tests.length
      const num_selected_tests = selected_tests.length
```

It is somewhat challenging to read this if-statement longer down: `num_test_suite == num_all_tests`.

File: bokehjs/test/devtools/devtools.ts [hunk 2]
```diff
@@ -752,6 +758,11 @@ async function run_tests(ctx: TestRunContext): Promise<boolean> {
       }
 
       if (failures != 0) {
+        if (num_selected_tests == num_all_tests) {
+          fail(`\n${chalk.cyan(failures)} of ${chalk.cyan(num_selected_tests)} tests ${chalk.red("failed")}`)
+        } else {
+          fail(`\n${chalk.cyan(failures)} of ${chalk.cyan(num_selected_tests)} selected tests ${chalk.red("failed")} (out of ${chalk.cyan(num_all_tests)} total tests)`)
+        }
```
Comment (bryevdv) [inline bokehjs/test/devtools/devtools.ts]: Maybe just provide all the numbers so any can be compared at a glance
```
6 passing, 7 failed of 13 selected tests (15 skipped out of 28 total tests)
```

Comment (hoxbro) [reply bokehjs/test/devtools/devtools.ts]: Maybe mimic pytest output?

image
image

Comment (bryevdv) [reply bokehjs/test/devtools/devtools.ts]: 👍 even better

Comment (mattpap) [reply bokehjs/test/devtools/devtools.ts]: The output is now:

image

**Review by bryevdv — APPROVED** [review]

**Review by hoxbro — APPROVED** [review]

**Pull request merged**